aten/src/ATen/native/cpu/GridSamplerKernel.cpp

```diff
@@ -0,0 +1,890 @@
+    );
+  }
+};
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ApplyGridSample ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Struct to apply grid sample (reading from input, interpolate, and write to
+// output).
+// See NOTE [ Grid Sample CPU Kernels ] for details.
+
+template<typename scalar_t>
+static inline void
+mask_scatter_add(const scalar_t *src, scalar_t* base_addr,
+                 const int_same_size_t<scalar_t> *offsets,
+                 const int_same_size_t<scalar_t> *mask, int64_t len) {
+  #pragma unroll
+  for (int64_t i = 0; i < len; i++) {
+    if (mask[i] & 0x01) {
+      base_addr[offsets[i]] += src[i];
+    }
+  }
+}
+
+template<typename scalar_t, int spatial_dim,
+         GridSamplerInterpolation interp,
+         GridSamplerPadding padding>
+struct ApplyGridSample;
+
+template<typename scalar_t, GridSamplerPadding padding>
+struct ApplyGridSample<scalar_t, 2, GridSamplerInterpolation::Bilinear, padding> {
+  using Vec = Vec256<scalar_t>;
+  using integer_t = int_same_size_t<scalar_t>;
+  using iVec = Vec256<integer_t>;
+
+  const int64_t inp_H;
+  const int64_t inp_W;
+  const int64_t inp_sH;
+  const int64_t inp_sW;
+  const int64_t C;
+  const int64_t inp_sC;
+  const ComputeLocation<scalar_t, padding> compute_H;
+  const ComputeLocation<scalar_t, padding> compute_W;
+  const bool must_in_bound = padding != GridSamplerPadding::Zeros;
+
+  ApplyGridSample(const TensorAccessor<scalar_t, 4>& input)
+    : inp_H(input.size(2))
+    , inp_W(input.size(3))
+    , inp_sH(input.stride(2))
+    , inp_sW(input.stride(3))
+    , C(input.size(1))
+    , inp_sC(input.stride(1))
+    , compute_H(input.size(2))
+    , compute_W(input.size(3)) {}
+
+  inline std::tuple<
+    Vec, Vec, Vec, Vec,       // distances to 4 sides
+    Vec, Vec, Vec, Vec,       // interpolation weights wrt 4 corners
+    Vec, Vec, Vec, Vec,       // in_bound masks
+    iVec, iVec                // y_n and x_w
+  >
+  compute_interp_params(const Vec& x, const Vec& y) const {
+    // get NE, NW, SE, SW pixel values from (x, y)
+    // assuming we get exact integer representation and just use scalar_t
+    // if we don't, the weights will be garbage anyways.
+    auto x_w = x.floor();
+    auto y_n = y.floor();
+
+    // get distances to each side
+    auto w = x - x_w;
+    auto e = Vec(1) - w;
+    auto n = y - y_n;
+    auto s = Vec(1) - n;
+
+    // get interpolation weights for each neighbor
+    // e.g., for the nw corder, the weight is `dist_to_south * dist_to_east`.
+    auto nw = s * e;
+    auto ne = s * w;
+    auto sw = n * e;
+    auto se = n * w;
+
+    auto i_x_w = convert_to_int_of_same_size(x_w);
+    auto i_y_n = convert_to_int_of_same_size(y_n);
+    auto i_x_e = i_x_w + iVec(1);
+    auto i_y_s = i_y_n + iVec(1);
+
+    // Use int comparison because it is much faster than float comp with AVX2
+    // (latency 1 cyc vs. 4 cyc on skylake)
+    // Avoid using the le and ge because those are not implemented in AVX2 and
```

Comment:
Not for this PR: but if le and ge have bad perf characteristics on AVX2, perhaps we shouldn't have overloads for them on the Vec256 class.

Reply:
(We can still provide a reference implementation of it, but under a name that makes it clear that it's simulated)